### ISSUE TITLE
Rules that require type resolution are not applied to KMP `commonMain`

### ISSUE THREAD

**Issue by kirillzh** [detekt/detekt#5961]
## Expected Behavior
Rules that require type resolution can be targeting `commonMain` source set out of the box.

## Current Behavior
Rules that require type resolution gets ignored for `commonMain` source set.
Steps:
1. Enable a rule that requires type resolution, for example `ForbiddenMethodCall`
2. In `commonMain` add code that breaks the rule
3. Enable build fail on rule violation
4. Enable detekt debugger: `debug = true` in plugin
5. Run `gradle detektMetadataCommonMain`
6. See that detekt doesn't fail, and see this warning printed out: `The rule 'ForbiddenMethodCall' requires type resolution but it was run without it.`

## Context
Considering that with KMP the idea is to share majority of the code between platforms through `commonMain` source set, I would expect Detekt to be able to run all rule targeting that source set, but that doesn't seem to be the case?

Relevant discussions:
- https://kotlinlang.slack.com/archives/C88E12QH4/p1639596467112800?thread_ts=1639594663.111700&cid=C88E12QH4
- https://kotlinlang.slack.com/archives/C88E12QH4/p1680055736056299?thread_ts=1679989098.335429&cid=C88E12QH4

**Comment by cortinico** [comment]
> Rules that require type resolution can be targeting `commonMain` source set out of the box.

As of today, we still haven't found a way to get Type Resolution working for non-JVM targets (JVM/Android). Therefore this is not possible as of today.
The current workaround is to use the Detekt Compiler Plugin.

**Comment by 3flex** [comment]
I don't think it will ever be possible to auto create a task that enables type resolution for the `commonMain` source set.

`commonMain` is a source set but it can't be compiled on its own. It's only used as a shared source set that is used in other compilations like `main` and `test`. For type resolution to work, detekt needs to analyse the result of a compilation, not a source set.

You can see the logical diagram here: https://kotlinlang.org/docs/multiplatform-discover-project.html#compilations

The plugin and docs need to be clearer about the distinction between source sets and compilations and what will and won't work for each.

**Comment by MattSkala** [comment]
The [documentation](https://detekt.dev/docs/1.23.8/gettingstarted/type-resolution#enabling-on-a-kmp-project) states that the `detektMetadataCommonMain` task “runs detekt with type resolution on the common module in the main source set”. Is this description inaccurate, or are we misunderstanding how it works?

**Comment by hqzxzwb** [comment]
> > Rules that require type resolution can be targeting `commonMain` source set out of the box.
> 
> As of today, we still haven't found a way to get Type Resolution working for non-JVM targets (JVM/Android). Therefore this is not possible as of today. The current workaround is to use the Detekt Compiler Plugin.

Now that [Kotlin analysis API](https://kotlin.github.io/analysis-api/index_md.html) is (almost?) stable, I think running type resolution on non-JVM platforms is possible now. Is there a plan for this? I have managed to build my own lint tool based on `analysis API` and migrated some detekt rules to it. The rules work correctly for non-JVM targets. Looking forward to migrating back to detekt.
